### PR TITLE
OSMD VexflowPatch: stave & stavesection

### LINKED ISSUE #1253: OSMD VexflowPatch: Stave
prevent a bug where a modifier width is NaN, leading to a VexFlow error
stave.setSection(section, y, xOffset = 0, fontSize = 12):
add xOffset, fontSize arguments (see stavesection.js)

### LINKED ISSUE #1256: OSMD VexflowPatch: StaveSection
stavesection.draw():
adjust rectangle positioning, make height depend on text height

### PR DESCRIPTION
**Pull request by rvilarl**
Fixes #1253
Fixes #1256 

@sschmidTU could you review this one?

@0xfe the visual regression does not show any difference although there is the difference bellow. Do you agree with this change? To me in makes more sense to use the text meassurement hieght and the question is also if we want to add a padding up / down

**Current**
Stave Multiple_Stave_Barline_Test Bravura
Stave Multiple_Stave_Barline_Test Gonville
Stave Multiple_Stave_Barline_Test Petaluma

**Reference**
Stave Multiple_Stave_Barline_Test Bravura
Stave Multiple_Stave_Barline_Test Gonville
Stave Multiple_Stave_Barline_Test Petaluma

### CODE REVIEW

**Comment by sschmidTU** [comment]
I would either set the default fontsize to 10, or not make the font bold. 12pt bold is too big and hard to read in the `Current` version.

@rvilarl by the way, your images don't have a white background, so they're almost unreadable in Github Dark Mode. I guess I have to use Light Mode for this.
image

**Comment by mscuthbert** [comment]
> @rvilarl by the way, your images don't have a white background, so they're almost unreadable in Github Dark Mode. I guess I have to use Light Mode for this.

Yeah, I keep thinking of putting a patch in that draws the images against a white background instead of transparent.  I realize that I haven't been reviewing visual diffs much because they're so hard for me to read.

**Comment by rvilarl** [comment]
> > @rvilarl by the way, your images don't have a white background, so they're almost unreadable in Github Dark Mode. I guess I have to use Light Mode for this.
> 
> Yeah, I keep thinking of putting a patch in that draws the images against a white background instead of transparent. I realize that I haven't been reviewing visual diffs much because they're so hard for me to read.

@sschmidTU @mscuthbert this is a good idea the generated png files are transparent which does not fit with the current dark trend.

**Comment by rvilarl** [comment]
The changes is no font independent. However the svg variants provide a height that is too large. I would appreciate a tip. I do not understand why `SVGtextelement.getbbox()` provides this excesive height more larger than `CanvasRenderingContext2D.measureText()`

image

**Comment by rvilarl** [comment]
> The changes is now font independent. However the svg tests provide a height that is too large. I do not understand why `SVGtextelement.getbbox()` provides this excesive height more larger than `CanvasRenderingContext2D.measureText()`. I would appreciate a tip. :)

@0xfe @ronyeh @AaronDavidNewman any idea?

**Comment by rvilarl** [comment]
**As disscused in #1270** 

> > You usually want max height and actual width with text, for placement. Because text has a baseline. you might want the actual height in some cases, like if you wanted to draw a glyph or a box around it.
> 
> What do you think that we should do here? Should `ctx.measureText(this.text).height` return the actual height of `this.text` or `textFormatter.maxHeight`

**And in this PR**
> The changes is now font independent. However the svg tests provide a height that is too large. I do not understand why `SVGtextelement.getbbox()` provides this excesive height larger than `CanvasRenderingContext2D.measureText()`. I would appreciate a tip. :)

@0xfe @ronyeh @AaronDavidNewman how should we proceed?

**Comment by ronyeh** [comment]
> `ctx.measureText(this.text).height` return the actual height of `this.text` or `textFormatter.maxHeight`

If I recall, the maxHeight is just a hard coded value. It used to be the height of the H character. But it turns out there are some lower case characters that are taller than H, like lowercase b. In any case, using maxHeight is just an approximation and provides an "upper bound" assuming the maxHeight value was correct. 

But ideally, a method called .measureText() should probably return the real height / width measurements, if it is possible to do so correctly and with good enough performance.

**Comment by rvilarl** [comment]
> But ideally, a method called .measureText() should probably return the real height / width measurements, if it is possible to do so correctly and with good enough performance.

There is no performance issue, the function is beying called anyway for the width. The only thing that is pending is to understand why the svg variant calculates the excesive height. :(

**Comment by rvilarl** [comment]
According to https://stackoverflow.com/questions/36146331/javascripts-getbbox-provides-a-box-that-is-taller-than-its-content-text-why

_The bounding box returned by getBBox() is not the tight bounding box around the glyphs. It includes the full em-box height of the text - including allowance for ascenders (tallest characters in font) and descenders (lowest below baseline)._

@0xfe @sschmidTU @ronyeh how do we proceed then?

File: src/svgcontext.ts [hunk 1]
```diff
@@ -91,13 +91,7 @@ class MeasureTextCache {
     const bbox = txt.getBBox();
     svg.removeChild(txt);
 
-    const fontSizeInPt: string = attributes['font-size'];
-
-    // Remove the trailing 'pt' from the font size and scale to convert from points to canvas pixel units.
-    // CSS specifies dpi to be 96 and there are 72 points to an inch: 96/72 == 4/3.
-    const height = Font.convertSizeToPixelValue(fontSizeInPt);
-
```
Comment (ronyeh) [inline src/svgcontext.ts:99]: What was the reasoning behind removing this chunk of code?

I ported this code in my Fonts PR, but I did not write the original "algorithm".... It is possible that the original author realized that the SVG bounding box was too high. It seems like the approach just uses the font size as a proxy, and turns it into a value in pixel units.

**Comment by ronyeh** [comment]
I spent some time looking around, and indeed the browser's SVG API does not provide a function to retrieve a tight bounding box.

It depends on what text we are sticking in the box. If the text has no descenders (the parts hanging at the bottom of the lowercase y, j, g) then we would need to reduce the height of the rectangle. Similarly, the height of the bbox leaves space for accent marks on upper case letters (Á, É, Ö, Ñ). So if our text has none of that, we'd also need to shorten the containing rectangle.

I put together a demo here:
https://jsfiddle.net/tsq6jxgL/

The basic idea is to change the **dominant-baseline** property of SVG text elements and check the bounding box's Y coordinate to programmatically determine the height of the text. 

Hanging

So maybe we can do this to a text element somewhere offscreen, and determine the height of the capital A? Note that this method ignores accent marks on upper case characters, and also ignores descenders.

I also have no idea how well this is supported across browsers! :-O

https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/dominant-baseline
https://en.wikipedia.org/wiki/Ascender_(typography)
https://en.wikipedia.org/wiki/Descender
https://en.wikipedia.org/wiki/Baseline_(typography)


Happy Holidays!

**Comment by rvilarl** [comment]
Hi Ron, thanks for confirming that this is the case. I have had an idea that I will push and comment later today. It is basically making canvas go the svg way.

**Comment by ronyeh** [comment]
What did you think of my JSfiddle demo?

Also, did you see my online comment above on the removed chunk of code?

**Comment by rvilarl** [comment]
> What did you think of my JSfiddle demo?
> 
> Also, did you see my inline comment above on the removed chunk of code?

yes I did! it is impressive that you managed to get the actual height using `'dominant-baseline` I tried with no success. But now and probably to find an alternative to the luck of success I thought that the opposite approach would have some benefits:
- If we have several sections with different texts, ti would be nice to have the rects aligned
- All the data to get the same approach in canvas was already there. There was no need for extra calculations.

I am almost there. I will push and then we decide if it looks fine or it is better to go for the tight approach. And now we know how to do it, it is in your JSfiddle.

**Comment by rvilarl** [comment]
Now done with svg approach also in canvas. I had to modify `TextMeasure` to include `x` & `y` in order to have the required informationfor the location of the rect.

image

The code is now also font size independent. I have included the following test to check it.

image

**Comment by rvilarl** [comment]
Ready to review

File: src/svgcontext.ts (hunk 1 above)
Comment (rvilarl) [reply src/svgcontext.ts:99]: This one was getting a fixed height rather than returning the ones calculated by `getBBox`

**Comment by rvilarl** [comment]
@0xfe I believe that this one is ready to merge, any concerns from your side?

**Comment by sschmidTU** [comment]
Fantastic fix, Rodrigo! Looks great! Being font size independent is great too.

**Comment by rvilarl** [comment]
@0xfe any concerns with this one? It is ready for review

**Comment by 0xfe** [comment]
Thanks for working through this, folks. @rvilarl, change looks great. For a separate PR, we probably should have some options in `measureText` for whether or not to include the ascender. I can see how that would be useful for things like annotations.

**Pull request merged**